src/net/spdy/spdy_session.cc

```diff
@@ -2172,11 +2184,30 @@ void SpdySession::EnqueueResetStreamFrame(spdy::SpdyStreamId stream_id,
   });
 
   DCHECK(buffered_spdy_framer_.get());
+  int padding_len = base::RandInt(48, 72) - spdy::kDataFrameMinimumSize;
   std::unique_ptr<spdy::SpdySerializedFrame> rst_frame(
-      buffered_spdy_framer_->CreateRstStream(stream_id, error_code));
+      buffered_spdy_framer_->CreateRstStream(stream_id, error_code,
+                                             padding_len));
+
+  auto buffer = std::make_unique<SpdyBuffer>(std::move(rst_frame));
+
+  DecreaseSendWindowSize(padding_len);
+  buffer->AddConsumeCallback(base::BindRepeating(
```

Comment:
This callback is strictly harmless but it can be elided because SpdySession::OnWriteBufferConsumed only takes effect when `consume_source == SpdyBuffer::DISCARD`, which only happens in `SpdyBuffer::~SpdyBuffer()`, which means some SpdyBuffer created in SpdySession::CreateDataBuffer and other places are destroyed before sent out. But in our case this strictly does not happen as we don't use SpdyBuffer.

Reply:
Sorry for this late reply, but I just realized that this callback is indeed necessary.

Thinking about that one party enqueues a RST_STREAM to the write queue, but before sending it on wire it receives a RST_STREAM from the other party, what will happen? `SpdyWriteQueue::RemovePendingWritesForStream` will be called, and then `SpdyBuffer::~SpdyBuffer()`. Without calling SpdySession::OnWriteBufferConsumed, the DISCARDed data frame will cause a leak in send window size. I've observed client and server sometimes send RST_STREAM to each other in a very tight time interval, so the situation I described above is possible.

Reply:
"this strictly does not happen as we don't use SpdyBuffer" isn't correct because there is still SpdyBuffer created downstream, therefore DISCARD might be invoked nonetheless.

I didn't look very deep, but isn't the rst frame created here associated with the session, not a stream? Why is RemovePendingWritesForStream involved here?

The RstStream frame is queued by SpdySession::EnqueueSessionWrite, which sets the stream to nullptr. When SpdyWriteQueue::RemovePendingWritesForStream is called the RstStream frame with nullptr stream is not erased from the queue, thus "and then SpdyBuffer::~SpdyBuffer()" is incorrect. From principle the RstStream frame is concerned with connection-level flow control not stream-level, so RemovePendingWrites**ForStream** shouldn't be an issue.

Reply:
I've checked it again. Yes, you were right.

The RST_STREAM is not associated with a stream, so not touched by RemovePendingWritesForStream. After receiving a RST_STREAM from server, it still sends on wire the RST_STREAM with the fake DATA anyway.

Interested by this behavior, I also checked the handling of receiving DATA on closed stream on server side. Golang updates session recv window properly. But haproxy doesn't. Haproxy just discards any frames on a closed stream, without checking whether it's a DATA and updating recv window. However, this bug is triggered by any DATA, not just padded ones.

Reply:
RFC has a MUST for this. I did report it to HAproxy https://www.mail-archive.com/haproxy@formilux.org/msg28780.html. I think the report was lost in the pile.

Haproxy's h2 stack has had a lot of changes and I didn't keep up to it. You seem to have a lot of energy then you can pursue this with Haproxy.

Reply:
If it is still a bug in Haproxy it may be because long-running connections are not a common use case for flow control leaks to have visible impact. You may need to stress this use case when reporting it.